pageserver/src/tenant/layer_map.rs

```diff
@@ -199,6 +216,7 @@ struct RangeSearchCollector<Iter>
 where
     Iter: Iterator<Item = (i128, Option<Arc<PersistentLayerDesc>>)>,
 {
+    in_memory_layer: Option<InMemoryLayerDesc>,
```

Comment:
Potential optimization for the future: given that in_memory_layer always cover the full key range, we don't need to consider the delta layers if they don't overlap. The good thing about the current approach is that the result from this iterator is equivalent to doing a `select_layer` on every possible key range in the layer map because we always evaluate delta+in-mem+image on each change point -- this sounds like the expected behavior of `ranged_search` test.

Reply:
Sorry, I'm not sure I understand. This currently works as follows:

We get one range to search. We then loop through all the "change points" in delta and image layer coverage while keeping track of what delta and image (if any) intersect the current section of the original range. For each section, we select a layer. The input to the layer selection is an option of each layer type according to what's currently in "scope".

> we don't need to consider the delta layers if they don't overlap

Are you saying we can just make the `(Some(delta), None, Some(image))` arm in `select_layer` `unreachable!`?